ProductiveRage.Immutable.Extensions/IEnumerableExtensions.cs

```diff
@@ -36,7 +36,10 @@ public static NonNullList<T> ToNonNullList<T>(this IEnumerable<T> values)
 			if (values == null)
 				throw new ArgumentNullException("values");
 
-			return new NonNullList<T>(values);
+			if (values.Any())
```

Comment:
Will this `.Any()` allocate an additional enumerator? (If so, this might result in more allocations than the empty lists, which are probably the minority case?)

Reply:
This will enumerate one item in the collection to check whether it has at least one element. I have no idea what the cost of that is.

For our eCMS Navigations work specifically, we have large tree hierarchies of pages that contain `NonNullList` of their nested child pages. The leaves of these trees will all have empty `NonNullList` objects, so this isn't a minority case for us here.

We transform records we get from the database to these nested lists using `ToNonNullList`. Having this extension method take care not to allocate new empty lists is helpful, as otherwise we'd have to perform this check on every call to `ToNonNullList` ourselves.

If the allocation cost of the additional enumerator is too much, do you know if there's a way to cache this enumerator so that the `Any` call can be made and if it returns true to pass in this cached object to a `NonNullList` constructor so that it won't allocate another new enumerator to populate all the items from?

Reply:
@dionrhys What type of collection are you passing in here? It's a T[] or List<T> then we could do a type check on them and then look at Length/Count to get a guaranteed allocation-free approach that still addresses your particular issue?

Reply:
In **most** of the cases they'll be `IEnumerable<T>` created from Linq expressions. Only rarely are they from `List<T>`.

I don't think there's any problem here. The enumerator allocated will be in scope for just the call to the `Any()` method, and I'd expect the JS garbage collector to sweep that away very quickly (like .NET's Gen 1) after that reference has gone. (Here's .NET's implementation by the way: http://referencesource.microsoft.com/#System.Core/System/Linq/Enumerable.cs,1279)

The amount of newly-allocated empty `NonNullList` that happen if this check doesn't occur will be quite a lot, and they will live for a long time.

Here are comparisons of heap usage from the `NonNullList` constructor without and with the `Any()` check with `ToNonNullList` highlighted in both.

**Without `Any()` check (115.2 KB):**
![image](https://user-images.githubusercontent.com/844245/33272113-1c5c38ac-d381-11e7-9e33-39eb36b0855b.png)

**With `Any()` check (32.9 KB):**
![image](https://user-images.githubusercontent.com/844245/33272117-22881980-d381-11e7-8e5c-fd2b8d887d00.png)

I've done some page loading performance captures for without and with the `Any()` check and the check does show up as adding time to the `ToNonNullList` call.

**Without `Any()` check (constructor = 11.2 ms):**
![image](https://user-images.githubusercontent.com/844245/33272834-570b759c-d383-11e7-8ad6-2e77cfaed974.png)

**With `Any()` check (any check = 14.2 ms, constructor = 6.3 ms):**
![image](https://user-images.githubusercontent.com/844245/33272883-80407340-d383-11e7-9641-5beee9f5f953.png)
![image](https://user-images.githubusercontent.com/844245/33272888-84373538-d383-11e7-8110-8591a9402c7d.png)

So the new check adds time, but also cuts down the time of `NonNullList` constructions.

The extra time doesn't seem that much of an impact to me (we are calling new code obviously), but it could be a case of death by a thousand cuts. I wanted to balance out the saved memory vs some more cpu time. I also didn't average this out over many runs because we need to actually work on the navigations stuff!

Maybe this is something that's more useful on the .NET side only where we understand the .NET garbage collector a lot better. But I think the .NET side (in the eCMS back end services) would benefit from Microsoft's own System.Collections.Immutable instead in that case since that side doesn't need any Bridge/.NET code sharing.

Reply:
Reading this comment reminded me that IEnumerable's `Count()` method has an optimisation for ICollection:

```cs
        public static int Count<TSource>(this IEnumerable<TSource> source) {
            if (source == null) throw Error.ArgumentNull("source");
            ICollection<TSource> collectionoft = source as ICollection<TSource>;
            if (collectionoft != null) return collectionoft.Count;
            ICollection collection = source as ICollection;
            if (collection != null) return collection.Count;
            int count = 0;
            using (IEnumerator<TSource> e = source.GetEnumerator()) {
                checked {
                    while (e.MoveNext()) count++;
                }
            }
            return count;
        }
```

I thought maybe it'd do the same for `Any()`, however no such luck! :(

```cs
        public static bool Any<TSource>(this IEnumerable<TSource> source) {
            if (source == null) throw Error.ArgumentNull("source");
            using (IEnumerator<TSource> e = source.GetEnumerator()) {
                if (e.MoveNext()) return true;
            }
            return false;
        }
```

(Of course, whether Bridge has the same implementation is another question).

I don't know the cost of doing those two casts (to `ICollection<T>` and `ICollection`), since the Count optimisation is about avoid enumerating the whole list rather than avoiding an allocation (probably why it's not implemented for `Any()`, which only checks the first item) but maybe something to test.